### PR TITLE
exchanges/request: abstract and consolidate rate limiting code to request package

### PR DESCRIPTION
**Pull request by shazbert**
# PR Description

Pulled from initial draft #1151 

* Allows context cancellation to occur while routine is being rate limited for any outbound REST request
* Reduces wrapper code size and bespoke implementations
* Fixes a couple of endpoints defaulting to unauth rates [BUG]

## Type of change

Please delete options that are not relevant and add an `x` in `[]` as item is complete.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How has this been tested

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration and
also consider improving test coverage whilst working on a certain feature or package.

- [ ] go test ./... -race
- [ ] golangci-lint run
- [ ] Test X

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation and regenerated documentation via the documentation tool
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally and on Github Actions with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: exchanges/bybit/ratelimit.go [hunk 1]
```diff
@@ -76,281 +68,67 @@ const (
...
-		GetSpotCrossMarginTradeLoanInfoRate:       request.NewRateLimit(time.Second, 50),
-		GetSpotCrossMarginTradeAccountRate:        request.NewRateLimit(time.Second, 50),
-		GetSpotCrossMarginTradeOrdersRate:         request.NewRateLimit(time.Second, 50),
-		GetSpotCrossMarginTradeRepayHistoryRate:   request.NewRateLimit(time.Second, 50),
-		SpotCrossMarginTradeLoanRate:              request.NewRateLimit(time.Second, 20),
-		SpotCrossMarginTradeRepayRate:             request.NewRateLimit(time.Second, 20),
-		SpotCrossMarginTradeSwitchRate:            request.NewRateLimit(time.Second, 20),
+func SetRateLimit() request.RateLimitDefinitions {
+	return request.RateLimitDefinitions{
+		defaultEPL:                               request.NewRateLimitWithToken(time.Second*5 /*See: https://bybit-exchange.github.io/docs/v5/rate-limit*/, 120, 1),
```
Comment (gloriousCode) [inline exchanges/bybit/ratelimit.go:74]: Looks like this has been upped to 600 according to your link. But double check me please

File: exchanges/request/limit.go [hunk 2]
```diff
@@ -26,23 +30,43 @@ const (
...
 // does not need to set endpoint functionality.
 type BasicLimit struct {
-	r *rate.Limiter
+	r *RateLimiterWithToken
 }
 
 // Limit executes a single rate limit set by NewRateLimit
-func (b *BasicLimit) Limit(ctx context.Context, _ EndpointLimit) error {
-	return b.r.Wait(ctx)
+func (b *BasicLimit) Limit(context.Context, EndpointLimit) (*RateLimiterWithToken, error) {
```
Comment (gloriousCode) [inline exchanges/request/limit.go:37]: Is there any need for these params?

Comment (gloriousCode) [inline exchanges/request/limit.go:37]: Is there a need to return err?

**Review by gloriousCode — CHANGES_REQUESTED** [review]
Cool! Only basic things that I could see.

File: exchanges/binance/ratelimit.go [hunk 3]
```diff
@@ -77,7 +74,7 @@ const (
 	cFuturesOrderbook100Rate
 	cFuturesOrderbook500Rate
 	cFuturesOrderbook1000Rate
-	cFuturesKline100Rate
+	// cFuturesKline100Rate
```
[the remark this answers is not in the document]
Comment (shazbert) [reply exchanges/binance/ratelimit.go:77]: Wasn't being used and was default rate

File: exchanges/request/limit.go (hunk 2 above)
Comment (shazbert) [reply exchanges/request/limit.go:37]: There is no need for even the struct. 👻 I shall burn it with fire. 🔥

File: exchanges/bitfinex/ratelimit.go [hunk 4]
```diff
@@ -175,81 +175,81 @@ const (
 // SetRateLimit returns the rate limit for the exchange
 func SetRateLimit() request.RateLimitDefinitions {
```
Comment (gloriousCode) [inline exchanges/bitfinex/ratelimit.go:176]: Let's rn all of them/revert my suggestion ;D

File: exchanges/binance/ratelimit_test.go [hunk 5]
```diff
@@ -64,13 +67,16 @@ func TestRateLimit_LimitStatic(t *testing.T) {
 		"All Price Changes": spotPriceChangeAllRate,
 		"All Orders":        spotAllOrdersRate,
 	}
+
+	rl, err := request.New("rateLimitTest2", http.DefaultClient, request.WithLimiter(GetRateLimits()))
```
Comment (gloriousCode) [inline exchanges/binance/ratelimit_test.go:71]:
```
2024-05-01T05:22:42.0693820Z --- FAIL: TestRateLimit_LimitStatic (0.00s)
2024-05-01T05:22:42.0694020Z     ratelimit_test.go:72: 
2024-05-01T05:22:42.0694940Z         	Error Trace:	/Users/runner/work/gocryptotrader/gocryptotrader/exchanges/binance/ratelimit_test.go:72
2024-05-01T05:22:42.0695210Z         	Error:      	Received unexpected error:
2024-05-01T05:22:42.0696000Z         	            	cannot set up a new requester for rateLimitTest2: cannot reuse http client
2024-05-01T05:22:42.0696270Z         	Test:       	TestRateLimit_LimitStatic
```
[From here](https://github.com/thrasher-corp/gocryptotrader/actions/runs/8905732198/job/24456856832?pr=1477#step:9:4494)

**Review by gloriousCode — CHANGES_REQUESTED** [review]
Thanks for making those changes!

**Review by gloriousCode — APPROVED** [review]
tACK! rate limiter still functioning
image
(rate limit at 1.991s with all pairs enabled)

**Review by gloriousCode — APPROVED** [review]
tACK deribit changes

**Review by thrasher- — APPROVED** [review]
Great stuff @shazbert

**Pull request merged**